File: src/main/java/service/elementum/android/ForegroundService.java
package service.elementum.android;

import android.net.Uri;
import android.os.Build;

import ilchev.stefan.binarywrapper.BaseForegroundService;

public class ForegroundService extends BaseForegroundService {

	@Override
	protected Class<?> getMainActivityClass() {
		return MainActivity.class;
	}

	@Override
	protected DaemonRunnable getDaemonRunnable(Uri data) {
		var fragment = data != null ? data.getFragment() : null;
		return new DaemonRunnable(this, MAIN_HANDLER, fragment != null ? fragment.split("\0") : null);
	}

	@Override
	protected String getVersionName(Uri data) {
		return MainActivity.getVersionName(data);
	}

	@Override
	protected String getUpdateFileName(String versionName) {
		var abi = Build.SUPPORTED_ABIS[0];
		return String.join("-", BuildConfig.PROJECT_NAME, abi, BuildConfig.BUILD_TYPE, versionName) + ".apk";
	}

	@Override
	protected Uri getUpdateDownloadUri(String versionName) {
		return Uri.parse(BuildConfig.REPO_URL + getUpdateFileName(versionName));
	}
}

File: ElementumService/src/main/kotlin/service/elementum/android/ForegroundService.kt
package service.elementum.android

import android.net.Uri
import android.os.Build
import ilchev.stefan.binarywrapper.BaseForegroundService

class ForegroundService : BaseForegroundService() {

	override val mainActivityClass = MainActivity::class.java

	override fun getDaemonInvoker(
		data: Uri?
	) = DaemonInvoker(this, mainHandler, *data?.fragment?.split("\u0000")?.toTypedArray() ?: emptyArray())

	override fun getVersionName(
		data: Uri?
	) = data?.schemeSpecificPart

	override fun getUpdateFileName(
		versionName: String
	) = "${BuildConfig.PROJECT_NAME}-${Build.SUPPORTED_ABIS[0]}-${BuildConfig.BUILD_TYPE}-$versionName.apk"

	override fun getUpdateDownloadUri(
		versionName: String
	): Uri = Uri.parse("${BuildConfig.REPO_URL}${getUpdateFileName(versionName)}")
}
